### DISCUSSION TITLE
Trivy IaC scanner crashed when run in container or github actions - 137 exit code

### DISCUSSION THREAD

**Discussion by obounaim** [aquasecurity/trivy#6517] [Bugs]
### Description

Trivy was OOM killed while scanning for Misconfiguration / IaC in Github Action or in Docker. Exit code 137.

`trivy config terraform/ --debug`

Tested with other commands, gave same behavior:
`trivy fs --scanners misconfig terraform/ --debug`
`trivy repo --scanners misconfig terraform/ --debug`

`du -sch terraform 
 12M    terraform
 12M    total`
 
 The scan works just fine on local system.

### Desired Behavior

Trivy misconfiguration complete successfully

### Actual Behavior

Trivy crashes with 137 exit code

### Reproduction Steps

```bash
A scan of an private Terraform repository for misconfiguration with Trivy
```

### Target

Filesystem

### Scanner

Misconfiguration

### Output Format

Table

### Mode

None

### Debug Output

```bash
2024-04-18T06:42:38.759Z        DEBUG   [misconf] 42:38.759982301 terraform.executor               Adapting modules...
2024-04-18T06:42:39.157Z        DEBUG   [misconf] 42:39.157767052 terraform.executor               Adapted 1 module(s) into defsec state data.
2024-04-18T06:42:39.157Z        DEBUG   [misconf] 42:39.157833385 terraform.executor               Using max routines of 11
2024-04-18T06:42:39.157Z        DEBUG   [misconf] 42:39.157841177 terraform.executor               Applying state modifier functions...
2024-04-18T06:42:39.157Z        DEBUG   [misconf] 42:39.157924468 terraform.executor               Initialized 486 rule(s).
2024-04-18T06:42:39.157Z        DEBUG   [misconf] 42:39.157943385 terraform.executor               Created pool with 11 worker(s) to apply rules.
2024-04-18T06:42:50.806Z        DEBUG   [misconf] 42:50.806293543 terraform.scanner.rego           Scanning 1 inputs...
Killed
4b401429f47a:/# echo $?
137
4b401429f47a:/#
```

### Operating System

Github Action

### Version

```bash
4b401429f47a:/# trivy --version
Version: 0.50.1
Policy Bundle:
  Digest: sha256:aa1640957b796d93a0ffc5d91237ee6b7ed9467b8f1825279384d29f91b9e590
  DownloadedAt: 2024-04-18 06:20:56.551859171 +0000 UTC
4b401429f47a:/#
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by nikpivkin** [comment]
Hi @obounaim !

How did you run Trivy in the docker?

**Reply by obounaim** [reply]
1. Using Github Action: https://github.com/aquasecurity/trivy-action

name: Security 
on:
  push:
    branches:
      - "*"
  pull_request:
  workflow_call:

jobs:
  security:
    name: security 
    runs-on: ubuntu-large
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
      - name: Run Trivy vulnerability scanner in IaC mode
        uses: aquasecurity/trivy-action@master
        with:
          scan-type: 'config'
          hide-progress: false

2. Locally using Github Action Docker Image

docker build .
docker run -v /path/terraform:/terraform --entrypoint bash -it 28495064ed63
trivy config terraform/ --debug

**Reply by obounaim** [reply]
Is there any switch or command that I can use to get more information why it is failing ? or to generate more logs

**Reply by nikpivkin** [reply]
You ran Trivy with the debug flag, which gives maximum information. How big is your repository? And is it private?

**Reply by obounaim** [reply]
Yes it is private. The repository is about 12MB in size.
When I run Trivy on my local machine memory consumption just goes crazy.

![Screenshot 2024-04-18 at 13 56 45](https://github.com/aquasecurity/trivy/assets/1151352/aefbf2cc-0218-45ca-a681-0824e250b654)

**Reply by nikpivkin** [reply]
There's some kind of leak. Since the repository is private, if I share the code, will you be able to run profiling and send me the results?

**Reply by obounaim** [reply]
yes, I should be able to do it.

**Reply by nikpivkin** [reply]
I've created a [repo](https://github.com/nikpivkin/trivy-pprof). Can you clone and run `go run ...`? Two files with profiling results will appear after the command is completed. It's just a wrapper over Trivy with profiling enabled.

**Reply by obounaim** [reply]
Do `heap.prof` and `cpu.prof` contain any private data ? I am not conformable sharing them publicly.

**Reply by nikpivkin** [reply]
I'll check

**Reply by nikpivkin** [reply]
Can you run the following commands and share the result?
```sh
go tool pprof heap.prof
top20
```

This will only show the memory used by the functions.

**Reply by obounaim** [reply]
trivy-pprof % go tool pprof heap.prof
File: trivy-pprof
Type: inuse_space
Time: Apr 18, 2024 at 3:06pm (CEST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20 
Showing nodes accounting for 8931.43MB, 98.64% of 9054.63MB total
Dropped 259 nodes (cum <= 45.27MB)
Showing top 20 nodes out of 44
      flat  flat%   sum%        cum   cum%
    3072MB 33.93% 33.93%     3072MB 33.93%  encoding/json.(*Decoder).refill
 2397.47MB 26.48% 60.41%  2672.98MB 29.52%  github.com/aquasecurity/trivy/pkg/iac/types.(*Metadata).ToRego
    2048MB 22.62% 83.02%     2048MB 22.62%  bytes.growSlice
 1056.63MB 11.67% 94.69%  1056.63MB 11.67%  encoding/json.Marshal
  246.02MB  2.72% 97.41%   276.02MB  3.05%  fmt.Sprintf
   73.03MB  0.81% 98.22%   104.99MB  1.16%  encoding/json.(*decodeState).objectInterface
   20.32MB  0.22% 98.44%  2709.31MB 29.92%  github.com/aquasecurity/trivy/pkg/iac/rego/convert.SliceToRego
      16MB  0.18% 98.62%  2688.99MB 29.70%  github.com/aquasecurity/trivy/pkg/iac/types.IntValue.ToRego
    1.96MB 0.022% 98.64%   101.99MB  1.13%  encoding/json.(*decodeState).arrayInterface
         0     0% 98.64%     2048MB 22.62%  bytes.(*Buffer).Write
         0     0% 98.64%     2048MB 22.62%  bytes.(*Buffer).grow
         0     0% 98.64%  3176.99MB 35.09%  encoding/json.(*Decoder).Decode
         0     0% 98.64%     3072MB 33.93%  encoding/json.(*Decoder).readValue
         0     0% 98.64%   104.99MB  1.16%  encoding/json.(*decodeState).object
         0     0% 98.64%   104.99MB  1.16%  encoding/json.(*decodeState).unmarshal
         0     0% 98.64%   104.99MB  1.16%  encoding/json.(*decodeState).value
         0     0% 98.64%   104.99MB  1.16%  encoding/json.(*decodeState).valueInterface
         0     0% 98.64%     2048MB 22.62%  encoding/json.(*encodeState).reflectValue
         0     0% 98.64%     2048MB 22.62%  encoding/json.arrayEncoder.encode
         0     0% 98.64%     2048MB 22.62%  encoding/json.interfaceEncoder

**Reply by nikpivkin** [reply]
@obounaim Thanks. 

No, these files do not contain sensitive data. The files resulting from heap profiling contain information about object addresses and sizes, as well as allocation stacks. CPU profiling files include a call graph and the percentage of CPU utilization for each function.

**Reply by obounaim** [reply]
@nikpivkin  do you need any more input from me ?

**Reply by simar7** [reply]
@obounaim Nikita is off this week. Once he's back he can assist you.

**Reply by simar7** [reply]
@obounaim Looks like this issue is reproducible in https://github.com/aquasecurity/trivy/discussions/6549 so I've made https://github.com/aquasecurity/trivy/issues/6557 to track it.

**Reply by nikpivkin** [reply]
@obounaim Yes, would you be able to send me the `heap.prof` and `mem.prof` files for investigation?

**Reply by obounaim** [reply]
@nikpivkin, I would like to share the full files. But because it was a scan of an internal project I would like to avoid any leak of internal code/information.

**Reply by obounaim** [reply]
Hi @nikpivkin , we are still facing the same Memory issue even with the latest trivy version.

**Reply by nikpivkin** [reply]
Hi @obounaim ! We recently solved a [problem](https://github.com/aquasecurity/trivy/issues/6557) with high memory usage, but apparently you have a different problem. I replied [above](https://github.com/aquasecurity/trivy/discussions/6517#discussioncomment-9162242) that dumps don't store sensitive data.

**Reply by simar7** [reply]
@obounaim please share the input with us so we can investigate.

**Reply by obounaim** [reply]
@nikpivkin , @simar7  where should I share the files ?
Is it okay to just upload them here ?

**Reply by nikpivkin** [reply]
@obounaim You can upload it here or to my e-mail

**Reply by obounaim** [reply]
[profile.zip](https://github.com/aquasecurity/trivy/files/15469546/profile.zip)
@nikpivkin, please find the files here attached.

**Reply by nikpivkin** [reply]
This profile is different from the one you had before. A brief excerpt shows that there was no leak.
```sh
Showing top 20 nodes out of 342
      flat  flat%   sum%        cum   cum%
   24.66MB 27.57% 27.57%    26.72MB 29.87%  github.com/aquasecurity/trivy/pkg/iac/scan.(*Result).GetCode
    5.56MB  6.21% 33.79%     5.56MB  6.21%  regexp/syntax.(*compiler).inst (inline)
       4MB  4.47% 38.26%        4MB  4.47%  github.com/open-policy-agent/opa/ast.(*state).Loc (inline)
    3.50MB  3.91% 42.17%        6MB  6.71%  encoding/json.(*decodeState).objectInterface
    3.01MB  3.36% 45.54%     3.01MB  3.36%  io.ReadAll
```

I've also updated this repository. https://github.com/nikpivkin/trivy-pprof

**Reply by obounaim** [reply]
It is better than last time but, it still uses 13GB for memory. Please find attached the new files.
[profile.zip](https://github.com/aquasecurity/trivy/files/15471784/profile.zip)

**Reply by simar7** [reply]
@obounaim it would be really helpful if you can send us an input that can help us reproduce the issue. 

You can reduce the input to a minimal reproducible example.

**Reply by obounaim** [reply]
I would love to, but the repository I am scanning contains a lot of files. I am not sure what input to give you.

**Reply by obounaim** [reply]
The issue seems to be related to GCP resources in terraform code.

**Reply by simar7** [reply]
>  I am not sure what input to give you.

Input that can help us reproduce this behavior as we have been unable to do so.

**Reply by obounaim** [reply]
Hi, I may have some new information on this topic. The high memory usage is caused by a `main.tf` file that manages 46 `google_compute_firewall` resource. Reducing the number of resources in the file reduces the memory usage proportionally.

**Reply by nikpivkin** [reply]
@obounaim Thanks, that helped find the problem.

**Comment by nikpivkin** [comment]
Track https://github.com/aquasecurity/trivy/issues/7548
